text/0000-newtypes.md

```diff
@@ -0,0 +1,263 @@
+If you get a `usize` for a userid, you can accidentally use it to index the
+`Vec<Pet>`. Since these ids have nothing in common, it might be desirable to
+make sure that you have a custom id type that cannot be confused with any other
+id type:
+
+```rust
+type UserIndex is new usize;
+// assume TVec is essentially `Vec` with a generic arg for the index type
+type UserVec is new TVec<UserIndex, User>;
+type PetIndex is new usize;
+type PetVec is new TVec<PetIndex, Pet>;
+
+fn foo(&mut self, u: UserIndex, p: PetIndex) {
+    self.users[p].pets.add(u); // ERROR users array can only be indexed by UserIndex
+    self.users[u].pets.add(p); // correct
+}
+```
+
+# Reference-level explanation
+
+Steal the `is new` syntax from Ada's newtypes by extending type aliases
+declarations with the `type` keyword.
+
+```rust
+type Inch is new usize;
+type Cm is new usize;
+
+// We want to do generic manipulations
+fn calc_distance<T: Sub>(start: T, end: T) -> T {
+    end - start
+}
+
+// Initialize the same way as the underlying types
+let (start_inch, end_inch): (Inch, Inch) = (10, 18);
+let (start_cm, end_cm): (Cm, Cm) = (2, 5);
+
+// Here `calc_distance` operates on the types `Inch` and `Cm`,
+// where previously we had to cast to and from `usize`.
+let inch_dist = calc_distance(start_inch, end_inch);
+let cm_dist = calc_distance(start_cm, end_cm);
+
+println!("dist: {} and {}", inch_dist, cm_dist);
+
+// Disallowed at compile time
+let not_allowed = calc_distance(start_inch, end_cm);
+```
+
+It would also allow generics:
+
+```rust
+struct A<N, M> { n: N, m: M }
+type B<T> is new A<usize, T>;
+
+let b = B { n: 2u, m: "this is a T" };
+```
+
+It would not be possible to use the newtype in place of the parent type,
+we would need to resort to traits.
+
+```rust
+fn bad(x: usize) { ... }
+fn good<T: Sub>(x: T) { ... }
+
+type Foo is new usize;
+let a: Foo = 2;
+bad(a); // Not allowed
+good(a); // Ok, Foo implements Sub
+```
+
+## Derived traits
+
+In the derived trait implementations the basetype will be replaced by the newtype.
+
+So for example as `usize` implements `Add<usize>`, `type Inch is new usize`
+would implement `Add<Inch>`.
+
+## Scoping
+
+Newtypes would follow the natural scoping rules:
+
+```rust
+type Inch is new usize; // Not accessible from outside the module
+pub type Cm is new usize; // Accessible
+
+use module::Inch; // Import into scope
+pub use module::Inch; // Re-export
+```
+
+### Reexporting private types
+
+Newtypes are allowed to be newtypes over private types:
+
+```rust
+mod foo {
+    struct Foo;
+    pub type Bar is new Foo;
+}
+let x: foo::Bar = ...; // OK
+let x: foo::Foo = ...; // Not OK
+```
+
+## Casting
+
+Newtypes can explicitly be converted to their base types, and vice versa.
+Implicit conversions are not allowed.
+This is achieved via the `Into` trait, since newtypes automatically implement
+`From<BaseType>` and `From<NewType> for BaseType`. In order to not expose new
```

Comment:
> This is achieved via the `Into` trait, since newtypes automatically implement `From<BaseType>` and `From<NewType> for BaseType`.

That sounds problematic. These `impl`s would be available to anyone as soon as `NewType` is exported due to the way `impl`s are not scoped in Rust. This means that you have no ability to write smart constructors that constrain `BaseType` further in the allowed set of values. While refinement types may partially (the expressiveness of refinement may/should not be turing complete, unlike unconstrained Rust-code, etc.) solve this in the future, we are far from having those currently. I think it is important that newtypes be allowed to provide only smart constructors such as `NewType::new(..) -> Self` that are user defined.

To that end, as an alternative I'd like to propose that the compiler generate a module:
```rust
mod NewType {
    pub fn from_base(BaseType) -> NewType {..}
    pub fn into_base(NewType) -> BaseType {..}
```
Another alternative may be inherent and associated methods on `NewType`.
The `From` impls should only be generated and iff the user does `derive(From, Into)` or perhaps `derive(Newtype)` (with a possible actual trait `Newtype`) to avoid conflicts with custom-derive.

Some notes discussing `T<NewType> -> T<BaseType>` where `T` may be `&`, `Box`, `Arc`, `Vec`, .. , would be nice even if you are proposing we not allow those conversions. In that case, notes on why the conversions should not be allowed.